File: ObjC/Person.m
//
//  Person.m
//  JavaToObjC
//
//  Created by Max Rogers on 2/10/14.
//  Copyright (c) 2014 Max Rogers. All rights reserved.
//

#import "Person.h"

@implementation Person
-(id) init:(NSString *)name :(NSDateComponents *)bday
{
    self = [super init];
    if(self){
        self.name = name;
        self.birthday = bday;
    }
    return self;
}

-(BOOL) isMale{return false;}
-(BOOL) isFemale{return false;}
-(NSString *) gender{return @"NONE";}

-(int) calculateAge{
    NSDateComponents *components = [[NSCalendar currentCalendar] components:NSDayCalendarUnit | NSMonthCalendarUnit | NSYearCalendarUnit fromDate:[NSDate date]];
    NSInteger year = [components year];
    return (int) (year - self.birthday.year);
}

-(NSString *) toString{
    NSString *result = [NSString stringWithFormat:@"\n%@:\t{Age:\t%d", self.name, [self calculateAge] ];
    result = [NSString stringWithFormat:@"%@\n\t\t\tBDay:\t%ld/%ld/%ld", result, self.birthday.month, self.birthday.day, (long)self.birthday.year ];
    if(self.spouse != nil){
        result = [NSString stringWithFormat:@"%@\n\t\t\tSpouse:\t%@", result, self.spouse.name ];
    }
    if(self.father != nil)
        result = [NSString stringWithFormat:@"%@\n\t\t\tFather:\t%@", result, self.father.name ];
    if(self.mother != nil)
        result = [NSString stringWithFormat:@"%@\n\t\t\tMother:\t%@", result, self.mother.name ];
    if(self.children != nil)
    {
        result = [NSString stringWithFormat:@"%@\n\t\t\tChildren:\t", result];
        for(Person *p in self.children) {
            result = [NSString stringWithFormat:@"%@\n\t\t\t\t%@", result, p.name ];
        }
    }
    result = [NSString stringWithFormat:@"%@}", result ];
    return result;
}

-(BOOL) alreadyClaimedChild:(Person *)p{
    for(Person *c in self.children)
    {
        if([p.name isEqualTo:c.name])
            return true;
    }
    return false;
}
-(void) addParents:(Person *)dad :(Person *)mom{
    self.father = dad;
    self.mother = mom;
}
-(void) addChild:(Person *)p{
    if(self.children == nil)
        self.children = [NSMutableArray array];
    [self.children addObject:p];
    if(self.isMale){
        p.father = self;
    }else{
        p.mother = self;
    }
    if(self.spouse != nil && ![self.spouse alreadyClaimedChild:p]){
        if(self.isMale){
            p.mother = self.spouse;
        }else{
            p.father = self.spouse;
        }
        [self.spouse addChild:p];
    }
}

+(NSDateComponents *) makeBday: (int) month : (int) day : (int) year{
    NSDateComponents *bday = [[NSDateComponents alloc] init];
    [bday setYear:year];
    [bday setMonth:month];
    [bday setDay:day];
    return bday;
}

-(void) setSpouse:(Person *)p{
    _spouse = p;
    if(p.spouse == nil){
        [p setSpouse:self];
    }
}
@end

File: Java/Person.java
import java.util.ArrayList;
import java.util.Calendar;

/**
 *	
 * @author max
 *
 */
public abstract class Person implements Gender
{
	private String name;
	private Calendar birthday;
	private Person father;
	private Person mother;
	private Person spouse;
	private ArrayList<Person> children;
	public Person(String name, Calendar birthday)
	{
		this.name = name;
		this.birthday = birthday;
	}
	public String getName(){return name;}
	public void setName(String name){ this.name = name;}
	public Calendar getBirthday(){return birthday;}
	public void setBirthday(Calendar birthday){ this.birthday = birthday;}
	public int calculateAge(){
		Calendar today = Calendar.getInstance();
		Calendar dob = birthday;
		int age = today.get(Calendar.YEAR) - dob.get(Calendar.YEAR);
		if (today.get(Calendar.MONTH) < dob.get(Calendar.MONTH)) {
			  age--;  
			} else if (today.get(Calendar.MONTH) == dob.get(Calendar.MONTH)
			    && today.get(Calendar.DAY_OF_MONTH) < dob.get(Calendar.DAY_OF_MONTH)) {
			  age--;  
			}
		return age;
	}
	
	public Person getFather(){ return father;}
	public void setFather(Person p){father = p;}
	public Person getMother(){ return mother;}
	public void setMother(Person p){mother = p;}
	public Person getSpouse(){return spouse;}
	public void setSpouse(Person p){
		spouse = p; 
		if(p.getSpouse() == null)
		{	p.setSpouse(this);	}
	}
	public void addChild(Person p){ 
		if(children == null)
			children = new ArrayList<Person>();
		children.add(p);
		if(isMale()){
			p.father = this;
		}else{
			p.mother = this;
		}
		if(spouse != null && !spouse.alreadyClaimedChild(p)){
			if(isMale()){
				p.mother = spouse;
				
			}else{
				p.father = spouse;
			}
			spouse.addChild(p);
		}
	}
	public boolean alreadyClaimedChild(Person p){
		if(children == null)
			return false;
		//normally i would setup some sort of search tree but since people rarely have more than 3 kids this shouldn't be an issue
		for(Person child : children)
		{
			if(child.name.equals(p.name))
				return true;
		}
		return false;
	}
	public void addParents(Person dad, Person mom){
		this.father = dad;
		this.mother = mom;
	}
	public ArrayList<Person> getChildren(){return children;}
	/**
	 * 
	 * @param int index i
	 * @return Person p
	 * @throws IndexOutOfBoundsException
	 */
	public Person getChild(int i) throws IndexOutOfBoundsException{
		return children.get(i);
	}
	
	public String toString(){
		String result =  name+":\t{Age:\t"+this.calculateAge(); 
		result+= "\n\t\tBDAY:\t"+birthday.get(Calendar.MONTH)+"/"+birthday.get(Calendar.DATE)+"/"+birthday.get(Calendar.YEAR);
		if(spouse != null)
			result += "\n\t\tSpouse:\t"+spouse.name;
		if(father != null)
			result += "\n\t\tFather:\t"+father.name;
		if(mother != null)
			result += "\n\t\tMother:\t"+mother.name;
		if(children != null){
			result += "\n\t\tChildren: ";
			for(int i=0;i<children.size();i++){
				result += "\n\t\t\t"+children.get(i).name;
			}
		}
		return result+"}\n";
	}
}
